internal/benchmarks/benchmark_test.go

```diff
@@ -0,0 +1,521 @@
+}
+
+func benchmarkGoogleapis_Protoparse(b *testing.B, factory func() *protoparse.Parser) {
+	par := runtime.GOMAXPROCS(-1)
+	cpus := runtime.NumCPU()
+	if par > cpus {
+		par = cpus
+	}
+	for i := 0; i < b.N; i++ {
+		// Buf currently batches files into chunks and then runs all chunks in parallel
+		chunks := make([][]string, par)
+		j := 0
+		total := 0
+		for ch := 0; ch < par; ch++ {
+			chunkStart := j
+			chunkEnd := (ch + 1) * len(googleapisSources) / par
+			chunks[ch] = googleapisSources[chunkStart:chunkEnd]
+			j = chunkEnd
+			total += len(chunks[ch])
+		}
+		require.Equal(b, total, len(googleapisSources))
+		var wg sync.WaitGroup
+		results := make([][]*desc.FileDescriptor, par)
+		errors := make([]error, par)
+		for ch, chunk := range chunks {
+			ch, chunk := ch, chunk
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				p := factory()
+				results[ch], errors[ch] = p.ParseFiles(chunk...)
+			}()
+		}
+		wg.Wait()
+		for _, err := range errors {
+			require.NoError(b, err)
+		}
+		var fdSet descriptorpb.FileDescriptorSet
+		fdSet.File = make([]*descriptorpb.FileDescriptorProto, 0, len(googleapisSources))
+		for _, chunk := range results {
+			for _, fd := range chunk {
+				fdSet.File = append(fdSet.File, fd.AsFileDescriptorProto())
+			}
+		}
+		writeToNull(b, &fdSet)
+	}
+}
+
+func BenchmarkGoogleapis_Protoc(b *testing.B) {
+	benchmarkGoogleapis_Protoc(b, "--include_source_info")
+}
+
+func BenchmarkGoogleapis_Protoc_NoSourceInfo(b *testing.B) {
+	benchmarkGoogleapis_Protoc(b)
+}
+
+func benchmarkGoogleapis_Protoc(b *testing.B, extraArgs ...string) {
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			args := make([]string, 0, len(googleapisSources)+5)
+			args = append(args, "-I", googleapisDir, "-o", nullOutputFile())
+			args = append(args, extraArgs...)
+			args = append(args, googleapisSources...)
+			cmd := exec.Command(protocPath, args...)
+			cmd.Stdin = nil
+			cmd.Stdout = nil
+			var errBuffer bytes.Buffer
+			cmd.Stderr = &errBuffer
+
+			err := cmd.Run()
+			if err != nil {
+				_, _ = os.Stderr.Write(errBuffer.Bytes())
+				b.Fatalf("failed to invoke protoc: %v", err)
+			}
+		}
+	})
+}
+
+func BenchmarkGoogleapis_Protocompile_SingleThreaded(b *testing.B) {
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			c := protocompile.Compiler{
+				Resolver: protocompile.WithStandardImports(&protocompile.SourceResolver{
+					ImportPaths: []string{googleapisDir},
+				}),
+				SourceInfoMode: protocompile.SourceInfoExtraComments,
+				// to really test performance compared to protoc and protoparse, we
+				// need to a single-threaded compile
+				MaxParallelism: 1,
+			}
+			fds, err := c.Compile(context.Background(), googleapisSources...)
+			require.NoError(b, err)
+			var fdSet descriptorpb.FileDescriptorSet
+			fdSet.File = make([]*descriptorpb.FileDescriptorProto, len(fds))
+			for i, fd := range fds {
+				fdSet.File[i] = protoutil.ProtoFromFileDescriptor(fd)
+			}
+			writeToNull(b, &fdSet)
```

Comment:
Lines 373-380 are the same as in other benchmarks in here. Is it worth extracting common code to a helper function or do you think that might impact benchmark results?

Reply:
Done. No noticeable impact to benchmark results.